core/audits/prioritize-lcp-image.js

```diff
@@ -47,42 +48,72 @@ class PrioritizeLcpImage extends Audit {
    *
    * @param {LH.Artifacts.NetworkRequest} request
    * @param {LH.Artifacts.NetworkRequest} mainResource
-   * @param {Array<LH.Gatherer.Simulation.GraphNode>} initiatorPath
+   * @param {InitiatorPath} initiatorPath
    * @return {boolean}
    */
   static shouldPreloadRequest(request, mainResource, initiatorPath) {
-    const mainResourceDepth = mainResource.redirects ? mainResource.redirects.length : 0;
-
     // If it's already preloaded, no need to recommend it.
     if (request.isLinkPreload) return false;
     // It's not a request loaded over the network, don't recommend it.
     if (NetworkRequest.isNonNetworkRequest(request)) return false;
     // It's already discoverable from the main document, don't recommend it.
-    if (initiatorPath.length <= mainResourceDepth) return false;
+    if (initiatorPath.length <= 2) return false;
     // Finally, return whether or not it belongs to the main frame
     return request.frameId === mainResource.frameId;
   }
 
   /**
    * @param {LH.Gatherer.Simulation.GraphNode} graph
    * @param {NetworkRequest} lcpRecord
-   * @return {{lcpNode: LH.Gatherer.Simulation.GraphNetworkNode|undefined, path: Array<LH.Gatherer.Simulation.GraphNetworkNode>|undefined}}
+   * @return {LH.Gatherer.Simulation.GraphNetworkNode|undefined}
    */
   static findLCPNode(graph, lcpRecord) {
-    let lcpNode;
-    let path;
-    graph.traverse((node, traversalPath) => {
-      if (node.type !== 'network') return;
+    for (const {node} of graph.traverseGenerator()) {
+      if (node.type !== 'network') continue;
       if (node.record.requestId === lcpRecord.requestId) {
-        lcpNode = node;
-        path =
-          traversalPath.slice(1).filter(initiator => initiator.type === 'network');
+        return node;
       }
-    });
-    return {
-      lcpNode,
-      path,
-    };
+    }
+  }
+
+  /**
+   * Get the initiator path starting with lcpRecord back to mainResource, inclusive.
+   * @param {NetworkRequest} lcpRecord
+   * @param {NetworkRequest} mainResource
+   * @return {InitiatorPath}
+   */
+  static getLcpInitiatorPath(lcpRecord, mainResource) {
+    /** @type {InitiatorPath} */
+    const initiatorPath = [];
+    let mainResourceReached = false;
+    /** @type {NetworkRequest|undefined} */
+    let request = lcpRecord;
+
+    while (request) {
+      mainResourceReached ||= request.requestId === mainResource.requestId;
```

Comment:
I think this variable can be scoped to the while loop, no?

Reply:
> I think this variable can be scoped to the while loop, no? 

no, it's `||=`

Reply:
but once it's true, it breaks. 🤷

Reply:
Ah yeah, I need to engage my brain harder next time. Originally this continued past the main resource to redirects, but it cuts off there now so the ||= isn't necessary and it can be scoped to the loop.

I'll update on the next change in here.